src/caf/distribute/furness.py

```diff
@@ -157,3 +161,234 @@ def doubly_constrained_furness(
+
+    idx_col:
+        Name of the columns in row_targets and col_targets that contain the
+        index data that matches seed_values index/columns
+
+    unique_col:
+        Name of the columns in row_targets and col_targets that contain the
+        values to target during the furness.
+
+    round_dp:
+        The number of decimal places to round the output values of the
+        furness to. Uses 4 by default.
+
+    unique_zones:
+        A list of unique zones to keep in the seed matrix when starting the
+        furness. The given productions and attractions will also be limited
+        to these zones as well.
+
+    unique_zones_join_fn:
+        The function to call on the column and index masks to join them for
+        the seed matrices. By default, a bitwise and is used. See pythons
+        builtin operator library for more options.
+
+    Returns
+    -------
+    furnessed_matrix:
+        The final furnessed matrix, in the same format as seed_values
+
+    completed_iters:
+        The number of completed iterations before exiting
+
+    achieved_rmse:
+        The Root Mean Squared Error difference achieved before exiting
+    """
+    # Init
+    row_targets = row_targets.copy()
+    col_targets = col_targets.copy()
+    seed_values = seed_values.copy()
+
+    row_targets = row_targets.reindex(columns=[idx_col, unique_col])
+    col_targets = col_targets.reindex(columns=[idx_col, unique_col])
+    row_targets = row_targets.set_index(idx_col)
+    col_targets = col_targets.set_index(idx_col)
+
+    # ## VALIDATE INPUTS ## #
+    ref_index = row_targets.index
+    if len(ref_index.difference(col_targets.index)) > 0:
+        raise ValueError("Row and Column target indexes do not match.")
+
+    if len(ref_index.difference(seed_values.index)) > 0:
+        raise ValueError("Row and Column target indexes do not match seed index.")
+
+    if len(ref_index.difference(seed_values.columns)) > 0:
+        raise ValueError("Row and Column target indexes do not match seed columns.")
+
+    assert_approx_equal(
+        row_targets[unique_col].sum(),
+        col_targets[unique_col].sum(),
+        err_msg="Row and Column target totals do not match. Cannot Furness.",
+    )
+
+    # ## TIDY AND INFILL SEED ## #
+    # Infill the 0 zones
+    seed_values = seed_values.mask(seed_values <= 0, seed_infill)
+    if normalise_seeds:
+        seed_values /= seed_values.sum().fillna(0)
+
+    # If we were given certain zones, make sure everything else is 0
+    if unique_zones is not None:
+        # Get the mask and extract the data
+        mask = pd_utils.get_wide_mask(
+            df=seed_values,
+            select=unique_zones,
+            join_fn=unique_zones_join_fn,
+        )
+        seed_values = seed_values.where(mask, 0)
+
+    # ## CONVERT TO NUMPY AND FURNESS ## #
+    row_targets = row_targets.values.flatten()
+    col_targets = col_targets.values.flatten()
+    seed_values = seed_values.values
```

Comment:
Use `to_numpy()` instead of values, I think pandas might have plans to change values in future
```suggestion
    row_targets = row_targets.to_numpy().flatten()
    col_targets = col_targets.to_numpy().flatten()
    seed_values = seed_values.to_numpy()
```